staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go

```diff
@@ -284,6 +289,20 @@ func (s *DefaultStorageFactory) NewConfig(groupResource schema.GroupResource) (*
 	}
 	glog.V(3).Infof("storing %v in %v, reading as %v from %#v", groupResource, codecConfig.StorageVersion, codecConfig.MemoryVersion, codecConfig.Config)
 
+	if storageConfig.Type == storagebackend.StorageTypeETCD3 {
+		if storageConfig.ResourcesToMonitor == nil {
+			storageConfig.ResourcesToMonitor = make(map[string]string)
+		}
+		resourcePrefix := s.ResourcePrefix(groupResource)
+		path := []string{"/registry"}
```

Comment:
this sets up a duplicate prefix construction path that doesn't honor the configured etcd prefix, and assumes nesting under the API group name. those aren't valid assumptions.

Reply:
Prefix is easily fixable, I'm not sure about the api group name - why it's not valid (except core/v1 ones)?

Reply:
none of the kube-apiserver types place themselves under group subpaths. see test/integration/etcd/etcd_storage_path_test.go for all the various non-standard key paths used.

I don't think we should set up another place like this where we try to pull together all the etcd prefix + resource prefix config values... if we want to run metrics on storage, why not do it inside generic store and storage, where this information already lives?

Reply:
I was looking into that yesterday, but got pulled into different things before I was able to get to the bottom of this. Short answer is that I have no idea how the storage code is structured.

Reply:
The way I understand it is that Storage (the layer that talks to etcd and the place where monitoring needs to live) is created before generic.Store(s) and get injected into Stores by *something* (Decorator)?.

So I can either modify the injection logic, to also pass some data down to Storage (which does have an unpleasant smell to it), or add a simple "Count" function to Storage. I'd rather do the latter, but I don't really want to make it full blown API-like request.

Reply:
The issue with the latter is that I have to either provide EXTREMELY inefficient implementations for etcdv2 and Cacher, or just always return -1/unimplemented error. Neither is particularly nice...

Reply:
> or add a simple "Count" function to Storage. I'd rather do the latter, but I don't really want to make it full blown API-like request

That is the most accurate representation of what this is attempting to do. If we want to pull info out of etcd in a new way, the Storage interface is the correct place to represent that.

> I have to either provide EXTREMELY inefficient implementations for etcdv2 and Cacher

Etcd2 returning an unimplemented error doesn't particularly bother me. Cacher could delegate to underlying store (as it does for most things), or add a `KeyCount()` function to the `cache.Store` interface

Reply:
This is what I'll try to send out today, assuming I'll stop getting interrupted...